oauth2_http/java/com/google/auth/oauth2/UserAuthorizer.java

```diff
@@ -285,61 +295,35 @@ public UserCredentials getCredentialsFromCode(String code, URI baseUri) throws I
-    }
-
-    UrlEncodedContent tokenContent = new UrlEncodedContent(tokenData);
-    HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
-    HttpRequest tokenRequest =
-        requestFactory.buildPostRequest(new GenericUrl(tokenServerUri), tokenContent);
-    tokenRequest.setParser(new JsonObjectParser(OAuth2Utils.JSON_FACTORY));
-
-    HttpResponse tokenResponse = tokenRequest.execute();
-
-    GenericJson parsedTokens = tokenResponse.parseAs(GenericJson.class);
-    String accessTokenValue =
-        OAuth2Utils.validateString(parsedTokens, "access_token", FETCH_TOKEN_ERROR);
-    int expiresInSecs = OAuth2Utils.validateInt32(parsedTokens, "expires_in", FETCH_TOKEN_ERROR);
-    Date expirationTime = new Date(new Date().getTime() + expiresInSecs * 1000);
-    String scopes =
-        OAuth2Utils.validateOptionalString(
-            parsedTokens, OAuth2Utils.TOKEN_RESPONSE_SCOPE, FETCH_TOKEN_ERROR);
-    AccessToken accessToken =
-        AccessToken.newBuilder()
-            .setExpirationTime(expirationTime)
-            .setTokenValue(accessTokenValue)
-            .setScopes(scopes)
-            .build();
-    String refreshToken =
-        OAuth2Utils.validateOptionalString(parsedTokens, "refresh_token", FETCH_TOKEN_ERROR);
-
+    TokenResponseWithConfig tokenResponseWithConfig =
+        getCredentialsFromCodeInternal(code, baseUri, additionalParameters);
     return UserCredentials.newBuilder()
-        .setClientId(clientId.getClientId())
-        .setClientSecret(clientId.getClientSecret())
-        .setRefreshToken(refreshToken)
-        .setAccessToken(accessToken)
-        .setHttpTransportFactory(transportFactory)
-        .setTokenServerUri(tokenServerUri)
+        .setClientId(tokenResponseWithConfig.getClientId())
+        .setClientSecret(tokenResponseWithConfig.getClientSecret())
+        .setAccessToken(tokenResponseWithConfig.getAccessToken())
+        .setRefreshToken(tokenResponseWithConfig.getRefreshToken())
+        .setHttpTransportFactory(tokenResponseWithConfig.getHttpTransportFactory())
+        .setTokenServerUri(tokenResponseWithConfig.getTokenServerUri())
         .build();
   }
 
+  /**
+   * Handles OAuth2 authorization code exchange and returns a {@link TokenResponseWithConfig} object
+   * containing the tokens and configuration details.
+   *
+   * @param code The authorization code received from the OAuth2 authorization server.
+   * @param callbackUri The URI to which the authorization server redirected the user after granting
+   *     authorization.
+   * @param additionalParameters Additional parameters to include in the token exchange request.
+   * @return A {@link TokenResponseWithConfig} object containing the access token, refresh token (if
+   *     granted), and configuration details used in the OAuth flow.
+   * @throws IOException If an error occurs during the token exchange process.
+   */
+  public TokenResponseWithConfig getTokenResponseFromAuthCodeExchange(
```

Comment:
qq, is it possible to have this return `UserCredentials` instead of `TokenResponseWithConfig`? I see from the above method that the TokenResponseWithConfig fields are mapped to a UserCredentials and wonder if the same can be done here.

I believe TokenResponseWithConfig is only exposed here and I would prefer not to expose a new class if possible

Reply:
Unfortunately it is not, as UserCredentials is specific to Google and has several methods that would not work. Specifically did it like this to enable support for Workforce Identity Federation.